### DISCUSSION TITLE
[Feature Request] Non-ephemeral runners

### DISCUSSION THREAD

**Discussion by blackliner** [runs-on/runs-on#190] [Ideas]
We have a bazel monorepo and need the stateful runners to avoid caching issues and speed up builds. Is there a plan to support this use-case?

I know it will add all kinds of new things to consider, like instance lifecycle management, scaling out/in... But Ephemeral runners are just too wasteful :-)

**Comment by crohr** [comment]
Hello, I don’t think non ephemeral runners will be supported. Have you tried the s3 backed caching yet?

**Reply by blackliner** [reply]
I have not tried out runs-on, I am exploring solutions right now. In the world of bazel, backing up 50GB of the sandbox via S3 is just a waste of resources, and you'd rather want to have this data "just waiting" for the next job to arrive.

It might be due to the scale of our repo, but this is also the main reason why most deployments based on kubernetes are not going to work out eventually, and we are going with ASG-like approaches.

**Comment by crohr** [comment]
Backing up 4GB (the latest test I did) takes about 20s, and that’s with a standard 2 CPU runner with EBS drive and only up to 12.5 Gbps bandwidth. You can probably get save/restore times for 50GB closer to 2min with larger runners and NVME disks?

Not saying this is ideal, but in terms of waste at least you don’t keep large runners around when you don’t need them.

**Reply by blackliner** [reply]
We have a continous stream of builds coming in, so for the right number of runners, there will always be a "healthy" queue of jobs waiting for one of the runners to free up.

To your numbers, keep in mind that these are possibly millions of files. So running the `tar` on that folder alone will take you a substantial amount of time, which can go into hours if you are using the default EBS root drive with 3k IOPS (and not leveraging the ephemeral instance storage).

TL&DR: I understand the advantages of ephemeral runners, but for some usecases, including ours, they are an antipattern.

**Reply by crohr** [reply]
Sure, I assume you would use instances with locally attached NVME disks (which RunsOn supports) if you require 50GB just for a cache anyway. And yes compression will take a while. I’d love to run the numbers though :)

But completely understand if having a stable pool sounds more reasonable. Will just trade for another set of issues with regards to security / cleaning / etc. Agree that an ASG based solution would probably work fine if the load is constant during work hours.

**Reply by blackliner** [reply]
Security, while important, is not a real problem in our use case. The runners are isolated, private org etc.
Cleanup, yes, but with a lifetimt of 24hrs, most issues are also not a real problem in daily life.

Good to hear you support NVMe drives, as I was not able to see that from the docs. Are you using it for the container data-root too? See
- https://github.com/buildkite/elastic-ci-stack-for-aws/blob/v6.30.0/packer/linux/conf/bin/bk-mount-instance-storage.sh
- https://github.com/buildkite/elastic-ci-stack-for-aws/blob/v6.30.0/packer/linux/conf/bin/bk-configure-docker.sh#L106-L113

**Reply by crohr** [reply]
Yes, docker data root is automatically mounted on the NVME drive(s) if drives are detected. If multiple drives, they are arranged as raid-0 array for faster speeds.

You might find the benchmark here interesting: https://runs-on.com/benchmarks/github-actions-runners/#disk-performance

I have to log off now (it’s late here in France) but happy to continue the discussion at a later time. Can also book some time in https://runs-on.com/call to discuss this further.

**Comment by crohr** [comment]
I'm going to close this for now. Please let me know if you tried any of the options here or went with another solution?
